mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala

```diff
@@ -0,0 +1,799 @@
+      if (row.isNullAt(ordinal)) {
+        null
+      } else {
+        row.getArray(ordinal).toDoubleArray
+      }
+    }
+
+    // Returns the array at a given index, or null if the array is null.
+    private def nullableArrayL(row: UnsafeRow, ordinal: Int): Array[Long] = {
+      if (row.isNullAt(ordinal)) {
+        null
+      } else {
+        row.getArray(ordinal).toLongArray
+      }
+    }
+
+    private def b(x: Array[Double]): BV[Double] = Vectors.dense(x).asBreeze
+
+    private def bl(x: Array[Long]): BV[Long] = BV.apply(x)
+
+    /**
+     * Sets the content of a buffer based on a single row (initialization).
+     *
+     * The buffer must be uninitialized first.
+     */
+    private def fillBufferWithRow(buffer: Buffer, v: Vector, w: Double): Unit = {
+      require(buffer.n == -1, (buffer.n, buffer))
+      val n = v.size
+      buffer.n = n
+      buffer.totalCount = 1L
+      buffer.totalWeightSum = w
+      buffer.totalWeightSquareSum = w * w
+
+      val arr = v.toArray
+      assert(arr.length == n, (arr.toSeq, n))
+      if (buffer.mean != null) {
+        buffer.mean = arr.clone()
+      }
+      if (buffer.m2n != null) {
+        buffer.m2n = Array.ofDim(n)
+      }
+      if (buffer.max != null) {
+        buffer.max = arr.clone()
+      }
+      if (buffer.min != null) {
+        buffer.min = arr.clone()
+      }
+
+      // The rest of these operations have efficient bulk versions.
+      v match {
+        case dv: DenseVector =>
+          if (buffer.m2 != null) {
+            buffer.m2 = Array.ofDim(n)
+            b(buffer.m2) := w * (b(arr) :* b(arr))
+          }
+          if (buffer.l1 != null) {
+            buffer.l1 = Array.ofDim(n)
+            b(buffer.l1) := numerics.abs(b(arr))
+          }
+
+        case sv: SparseVector =>
+          if (buffer.m2 != null) {
+            buffer.m2 = Array.ofDim(n)
+            v.foreachActive { (index, value) =>
+              buffer.m2(index) = w * value * value
+            }
+          }
+
+          if (buffer.l1 != null) {
+            buffer.l1 = Array.ofDim(n)
+            v.foreachActive { (index, value) =>
+              buffer.l1(index) = w * math.abs(value)
+            }
+          }
+      }
+
+      // In the case of the weightSum and NNZ, we also have to account for the value of
+      // the elements.
+      // TODO It would be nice to vectorize these operations too.
+      if (buffer.weightSum != null) {
+        buffer.weightSum = Array.ofDim(n)
+        v.foreachActive { (index, value) =>
+          if (value != 0.0) {
+            buffer.weightSum(index) = w
+          }
+        }
+      }
+
+      if (buffer.nnz != null) {
+        buffer.nnz = Array.ofDim(n)
+        v.foreachActive { (index, value) =>
+          if (value != 0.0) {
+            buffer.nnz(index) = 1L
+          }
+        }
+      }
+    }
+
+    private def updateInPlaceDense(buffer: Buffer, v: DenseVector, w: Double): Unit = {
+      val epsi = Double.MinPositiveValue
```

Comment:
I think the intention here is to make sequential operation conveniently and efficiently. AFAIK, ```cForRange``` is also very efficient. Thanks.